lib/ansible/modules/cloud/smartos/imgadm.py

```diff
@@ -0,0 +1,302 @@
+  imgadm:
+    uuid: '70e3ae72-96b6-11e6-9056-9737fd4d0764'
+    state: updated
+
+- name: Add a source
+  imgadm:
+    source: 'https://datasets.project-fifo.net'
+    state: present
+
+- name: Add a Docker source
+  imgadm:
+    source: 'https://docker.io'
+    type: docker
+    state: present
+
+- name: Remove a source
+  imgadm:
+    source: 'https://docker.io'
+    state: absent
+'''
+
+import re
+
+# Shortcut for the imgadm(1M) command. While imgadm(1M) supports a
+# -E option to return any errors in JSON, the generated JSON does not play well
+# with the JSON parsers of Python. The returned message contains '\n' as part of
+# the stacktrace, which breaks the parsers.
+IMGADM = 'imgadm'
+
+# Helper method to massage stderr
+def errmsg(stderr):
+    match = re.match('^imgadm .*?: error \(\w+\): (.*): .*', stderr)
+    if match:
+        return match.groups()[0]
+    else:
+        return 'Unexpected failure'
+
+def update_images(module):
+    uuid = module.params['uuid']
+    cmd = IMGADM + ' update'
+
+    if uuid != '*':
+        cmd += ' {}'.format(uuid)
+
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    # There is no feedback from imgadm(1M) to determine if anything
+    # was actually changed. So treat this as an 'always-changes' operation.
+    # Note that 'imgadm -v' produces unparseable JSON...
+    return rc, stdout, errmsg(stderr), True
+
+def manage_sources(module, present):
+    force = module.params['force']
+    source = module.params['source']
+    type = module.params['type']
+
+    cmd = IMGADM + ' sources'
+
+    if force:
+        cmd += ' -f'
+
+    if present:
+        cmd += ' -a {0} -t {1}'.format(source, type)
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        # Check the various responses.
+        # Note that trying to add a source with the wrong type is handled
+        # above as it results in a non-zero status.
+        changed = True
+
+        regex = 'Already have "{0}" image source "{1}", no change'.format(type, source)
+        if re.match(regex, stdout):
+            changed = False
+
+        regex = 'Added "%s" image source "%s"' % (type, source)
+        if re.match(regex, stdout):
+            changed = True
+
+        # Fallthrough, assume changes
+        return (rc, stdout, errmsg(stderr), changed)
+    else:
+        # Type is ignored by imgadm(1M) here
+        cmd += ' -d %s' % (source)
+        (rc, stdout, stderr) = module.run_command(cmd)
+
+        changed = True
+
+        regex = 'Do not have image source "%s", no change' % (source)
+        if re.match(regex, stdout):
+            changed = False
+
+        regex = 'Deleted ".*" image source "%s"' % (source)
+        if re.match(regex, stdout):
+            changed = True
+
+        return (rc, stdout, errmsg(stderr), changed)
+
+def manage_images(module, present):
+    uuid = module.params['uuid']
+    pool = module.params['pool']
+    state = module.params['state']
+
+    if state == 'vacumed':
```

Comment:
note, the spelling of this word is: vacuumed.  Probably should change it everywhere.